File: cmd/server/routes/discovery.go
package routes

import (
	"net/http"
	"os/exec"
	"strings"

	"kubelens/server/store"
)

func registerDiscovery(mux *http.ServeMux) {
	mux.HandleFunc("GET /api/api-resources", handleApiResources)
}

// ApiResource is one row of `kubectl api-resources`. JSON tags must match what
// api/utils/api-resources.ts emits — the frontend reads this route from either
// backend and cannot tell them apart.
type ApiResource struct {
	Name         string `json:"name"`         // bare plural, e.g. "virtualservices"
	Kind         string `json:"kind"`         // Kind, e.g. "VirtualService"
	Group        string `json:"group"`        // API group ("" = core)
	ResourceType string `json:"resourceType"` // kubectl target, e.g. "virtualservices.networking.istio.io"
}

// parseApiResources reads the table `kubectl api-resources` prints. There is no
// JSON output mode. Columns are NAME [SHORTNAMES] APIVERSION NAMESPACED KIND, and
// SHORTNAMES is optional, so read from the right: kind last, apiVersion last-2.
func parseApiResources(stdout string) []ApiResource {
	out := []ApiResource{}
	lines := strings.Split(stdout, "\n")
	if len(lines) > 0 {
		lines = lines[1:] // drop the header
	}
	for _, raw := range lines {
		line := strings.TrimSpace(raw)
		if line == "" {
			continue
		}
		cols := strings.Fields(line)
		if len(cols) < 4 {
			continue
		}
		name := cols[0]
		kind := cols[len(cols)-1]
		apiVersion := cols[len(cols)-3]
		group := ""
		if i := strings.Index(apiVersion, "/"); i >= 0 {
			group = apiVersion[:i]
		}
		resourceType := name
		if group != "" {
			resourceType = name + "." + group
		}
		out = append(out, ApiResource{Name: name, Kind: kind, Group: group, ResourceType: resourceType})
	}
	return out
}

func filterExcluded(resources []ApiResource) []ApiResource {
	groups, names := store.LoadDiscoveryExclude()
	exGroups := make(map[string]bool, len(groups))
	for _, g := range groups {
		exGroups[g] = true
	}
	exNames := make(map[string]bool, len(names))
	for _, n := range names {
		exNames[n] = true
	}
	kept := []ApiResource{}
	for _, r := range resources {
		if exGroups[r.Group] || exNames[r.Name] {
			continue
		}
		kept = append(kept, r)
	}
	return kept
}

// GET /api/api-resources — namespaced kinds the cluster actually has, for the
// visibility panel. A broken APIService (a metrics adapter that is down, say)
// makes kubectl exit non-zero while still listing the working resources on
// stdout, so salvage those instead of returning nothing.
func handleApiResources(w http.ResponseWriter, r *http.Request) {
	cmd := exec.Command("kubectl", "api-resources", "--verbs=list", "--namespaced=true")
	var stderr strings.Builder
	cmd.Stderr = &stderr
	stdout, err := cmd.Output()

	if err == nil {
		writeJSON(w, http.StatusOK, map[string]any{"resources": filterExcluded(parseApiResources(string(stdout)))})
		return
	}
	if len(stdout) > 0 {
		writeJSON(w, http.StatusOK, map[string]any{
			"resources": filterExcluded(parseApiResources(string(stdout))),
			"warning":   strings.TrimSpace(stderr.String()),
		})
		return
	}
	msg := strings.TrimSpace(stderr.String())
	if msg == "" {
		msg = err.Error()
	}
	writeJSON(w, http.StatusOK, map[string]any{"resources": []ApiResource{}, "error": msg})
}

File: api/routes/discovery.ts
import express from 'express';
import type { Request, Response } from 'express';
import { execFile } from 'child_process';
import { promisify } from 'util';
import { getDiscoveryExclude } from '../utils/config-loader';
import { parseApiResources } from '../utils/api-resources';

const execFileAsync = promisify(execFile);

const router = express.Router();

/** A failed execFile carries the child's output on the error object. */
interface ExecError extends Error {
  stdout?: string;
  stderr?: string;
}

// Parse the table, then drop the curated noise groups/resources from config.
function parseAndFilter(stdout: string) {
  const { groups, resources } = getDiscoveryExclude();
  const exGroups = new Set(groups);
  const exResources = new Set(resources);
  return parseApiResources(stdout).filter(r => !exGroups.has(r.group) && !exResources.has(r.name));
}

// GET /api/api-resources — namespaced kinds the cluster actually has, for the
// visibility panel. `kubectl api-resources` has no JSON output, so parse the table.
// Broken APIServices (e.g. a down metrics adapter) make kubectl exit non-zero but
// it still lists the working resources on stdout — salvage those.
router.get('/api-resources', async (req: Request, res: Response) => {
  try {
    const { stdout } = await execFileAsync(
      'kubectl', ['api-resources', '--verbs=list', '--namespaced=true'],
      { timeout: 30000, maxBuffer: 10 * 1024 * 1024 }
    );
    res.json({ resources: parseAndFilter(stdout) });
  } catch (err) {
    const e = err as ExecError;
    if (e.stdout) {
      res.json({ resources: parseAndFilter(e.stdout), warning: (e.stderr || '').trim() });
    } else {
      res.json({ resources: [], error: (e.stderr || '').trim() || e.message });
    }
  }
});

export = router;
